examples/dev_your_plugin/plugins/your_plugin.go

```diff
@@ -58,9 +60,44 @@ type filter struct {
 func (f *filter) DecodeHeaders(headers api.RequestHeaderMap, endStream bool) api.ResultAction {
 	hdr := http.Header{}
 	hdr.Set("content-type", "text/plain")
+	allHeaders := headers.GetAllHeaders()
+	msg := "Headers:\n"
+	for k, v := range allHeaders {
+		msg += k + ": " + strings.Join(v, ",") + "\n"
+	}
 	return &api.LocalResponse{
 		Code:   200,
-		Msg:    "Your plugin is run\n",
+		Msg:    "Your plugin is running\n" + msg,
 		Header: hdr,
 	}
 }
+
+// DecodeData might be called multiple times during handling the request body.
+// The endStream is true when handling the last piece of the body.
+func (f *filter) DecodeData(data api.BufferInstance, endStream bool) api.ResultAction {
+	if !endStream {
+		return api.WaitAllData
```

Comment:
We use DecodeRequest to handle body which needs to be fully buffered, see https://github.com/mosn/htnn/blob/main/site/content/zh-hans/docs/developer-guide/plugin_development.md#filter-manager-%E8%AE%BE%E8%AE%A1.

So `api.WaitAllData` is only allowed when processing headers. There will be an error log for it.

Reply:
Thanks! I see, so "DecodeRequest" is designed for buffered request while "DecodeData" is for streaming API?

Is it the same for "EncodeRequest" and "EncodeData"? @spacewander

Reply:
what if "DecodeRequest" and "DecodeData" co-exists in the plugin?

Reply:
> Is it the same for "EncodeRequest" and "EncodeData"?

Yes